[Bug] [TMrgSort] Format2场景，编译报错：error: 'pto.tmrgsort' op format2 expects dst/tmp shapes to match

### Component

PTO Dialect / ODS (include/PTO/IR)

### Description

**PR地址：**
https://github.com/mouliangyu/PTOAS/pull/262

**运行st：**
python test/tilelang_st/script/run_st.py -r sim -v a5 -t tmrgsort

**报错信息：**
loc("/home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/testcase/tmrgsort/tmrgsort.pto":481:5):
  error: 'pto.tmrgsort' op format2 expects dst/tmp shapes to match

<img width="1404" height="691" alt="Image" src="https://github.com/user-attachments/assets/0213e991-a73a-4f14-8a8c-221881fc1e60" />

### Reproduction (minimal)

```markdown
test/basic/expand_tile_op_tilelang_tmrgsort.pto

module {
  // Format2: 2-list merge sort (f32_2list_b64_basic)
  func.func @TMRGSORT_f32_2list_b64_basic(%src0_ptr: !pto.ptr<f32>, %src1_ptr: !pto.ptr<f32>, 
                                          %dst_ptr: !pto.ptr<f32>, %ex_ptr: !pto.ptr<i16>) {
    %c0   = arith.constant 0   : index
    %c1   = arith.constant 1   : index
    %c256 = arith.constant 256 : index
    %c512 = arith.constant 512 : index
    %c0_i32 = arith.constant 0 : i32

    %src0_view = pto.make_tensor_view %src0_ptr,
      shape = [%c1, %c1, %c1, %c1, %c256],
      strides = [%c256, %c256, %c256, %c256, %c1]
      : !pto.tensor_view<1x1x1x1x256xf32>
    %src1_view = pto.make_tensor_view %src1_ptr,
      shape = [%c1, %c1, %c1, %c1, %c256],
      strides = [%c256, %c256, %c256, %c256, %c1]
      : !pto.tensor_view<1x1x1x1x256xf32>
    %dst_view = pto.make_tensor_view %dst_ptr,
      shape = [%c1, %c1, %c1, %c1, %c256],
      strides = [%c512, %c512, %c512, %c512, %c1]
      : !pto.tensor_view<1x1x1x1x256xf32>

    %src0_part = pto.partition_view %src0_view,
      offsets = [%c0, %c0, %c0, %c0, %c0],
      sizes = [%c1, %c1, %c1, %c1, %c256]
      : !pto.tensor_view<1x1x1x1x256xf32> -> !pto.partition_tensor_view<1x1x1x1x256xf32>
    %src1_part = pto.partition_view %src1_view,
      offsets = [%c0, %c0, %c0, %c0, %c0],
      sizes = [%c1, %c1, %c1, %c1, %c256]
      : !pto.tensor_view<1x1x1x1x256xf32> -> !pto.partition_tensor_view<1x1x1x1x256xf32>
    %dst_part = pto.partition_view %dst_view,
      offsets = [%c0, %c0, %c0, %c0, %c0],
      sizes = [%c1, %c1, %c1, %c1, %c256]
      : !pto.tensor_view<1x1x1x1x256xf32> -> !pto.partition_tensor_view<1x1x1x1x256xf32>

    %src0_tile = pto.alloc_tile
      : !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=256, v_row=1, v_col=256,
                      blayout=row_major, slayout=none_box, fractal=512, pad=0>
    %src1_tile = pto.alloc_tile
      : !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=256, v_row=1, v_col=256,
                      blayout=row_major, slayout=none_box, fractal=512, pad=0>
    %dst_tile = pto.alloc_tile
      : !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=256, v_row=1, v_col=256,
                      blayout=row_major, slayout=none_box, fractal=512, pad=0>
    %tmp_tile = pto.alloc_tile
      : !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=512, v_row=1, v_col=512,
                      blayout=row_major, slayout=none_box, fractal=512, pad=0>
    %ex_vec = arith.constant dense<0> : vector<4xi16>

    pto.tload ins(%src0_part : !pto.partition_tensor_view<1x1x1x1x256xf32>)
              outs(%src0_tile : !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=256, v_row=1, v_col=256,
                                      blayout=row_major, slayout=none_box, fractal=512, pad=0>)
    pto.tload ins(%src1_part : !pto.partition_tensor_view<1x1x1x1x256xf32>)
              outs(%src1_tile : !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=256, v_row=1, v_col=256,
                                      blayout=row_major, slayout=none_box, fractal=512, pad=0>)

    pto.tmrgsort ins(%src0_tile, %src1_tile, %tmp_tile {exhausted = false} :
                    !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=256, v_row=1, v_col=256,
                                  blayout=row_major, slayout=none_box, fractal=512, pad=0>,
                    !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=256, v_row=1, v_col=256,
                                  blayout=row_major, slayout=none_box, fractal=512, pad=0>,
                    !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=512, v_row=1, v_col=512,
                                  blayout=row_major, slayout=none_box, fractal=512, pad=0>)
               outs(%dst_tile, %ex_vec :
                    !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=256, v_row=1, v_col=256,
                                  blayout=row_major, slayout=none_box, fractal=512, pad=0>,
                    vector<4xi16>)

    pto.tstore ins(%dst_tile : !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=256, v_row=1, v_col=256,
                                      blayout=row_major, slayout=none_box, fractal=512, pad=0>)
               outs(%dst_part : !pto.partition_tensor_view<1x1x1x1x256xf32>)
    return
  }
}


执行命令

  pto_file=test/basic/expand_tile_op_tilelang_tmrgsort.pto
  ptoas $pto_file \
    --pto-arch=a5 \
    --pto-backend=vpto \
    --enable-tile-op-expand \
    --vpto-emit-hivm-llvm \
    -o - \
    > tmp.ll \
    2> tmp.mlir
```

### Expected behavior

编译通过

### Actual behavior / error logs

```text
[INFO] ptoas: /home/caojian/workspace/git/github/mouliangyu/PTOAS/build/tools/ptoas/ptoas
target_dir: /home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st
run env shell: /usr/local/CANN/cann-9.0.0-beta.1/bin/setenv.bash
clean build: build
build failed: /usr/bin/cmake -S/home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st -B/home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/build --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start /home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/build/CMakeFiles /home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/build//CMakeFiles/progress.marks
make  -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/build'
make  -f testcase/tmrgsort/CMakeFiles/tmrgsort_kernel.dir/build.make testcase/tmrgsort/CMakeFiles/tmrgsort_kernel.dir/depend
make[2]: Entering directory '/home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/build'
[ 14%] ptoas: tmrgsort.pto -> tmrgsort_kernel.ll
cd /home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/build/testcase/tmrgsort && /usr/bin/cmake -DPTOAS_BIN=/home/caojian/workspace/git/github/mouliangyu/PTOAS/build/tools/ptoas/ptoas -DPTO_SRC=/home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/testcase/tmrgsort/tmrgsort.pto -DKERNEL_LL=/home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/build/testcase/tmrgsort/tmrgsort_kernel.ll -DPTOAS_ENABLE_INSERT_SYNC=ON -DPTOAS_PTO_LEVEL= -P /home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/testcase/run_ptoas_to_file.cmake
CMake Error at /home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/testcase/run_ptoas_to_file.cmake:54 (message):
  ptoas failed while generating
  /home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/build/testcase/tmrgsort/tmrgsort_kernel.ll:



  loc("/home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/testcase/tmrgsort/tmrgsort.pto":481:5):
  error: 'pto.tmrgsort' op format2 expects dst/tmp shapes to match


  loc("/home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/testcase/tmrgsort/tmrgsort.pto":555:5):
  error: 'pto.tmrgsort' op format2 expects dst/tmp shapes to match


  loc("/home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/testcase/tmrgsort/tmrgsort.pto":632:5):
  error: 'pto.tmrgsort' op format2 expects dst/tmp shapes to match


  loc("/home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/testcase/tmrgsort/tmrgsort.pto":725:5):
  error: 'pto.tmrgsort' op format2 expects dst/tmp shapes to match


  loc("/home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/testcase/tmrgsort/tmrgsort.pto":835:5):
  error: 'pto.tmrgsort' op format2 expects dst/tmp shapes to match


  loc("/home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/testcase/tmrgsort/tmrgsort.pto":947:5):
  error: 'pto.tmrgsort' op format2 expects dst/tmp shapes to match


  loc("/home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/testcase/tmrgsort/tmrgsort.pto":1045:5):
  error: 'pto.tmrgsort' op format2 expects dst/tmp shapes to match


  loc("/home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/testcase/tmrgsort/tmrgsort.pto":1139:5):
  error: 'pto.tmrgsort' op format2 expects dst/tmp shapes to match

  Error: Failed to parse MLIR.


make[2]: *** [testcase/tmrgsort/CMakeFiles/tmrgsort_kernel.dir/build.make:85: testcase/tmrgsort/tmrgsort_kernel.ll] Error 1
make[2]: Leaving directory '/home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/build'
make[1]: *** [CMakeFiles/Makefile2:114: testcase/tmrgsort/CMakeFiles/tmrgsort_kernel.dir/all] Error 2
make[1]: Leaving directory '/home/caojian/workspace/git/github/mouliangyu/PTOAS/test/tilelang_st/npu/a5/src/st/build'
make: *** [Makefile:91: all] Error 2

run failed: Command '['make', 'VERBOSE=1', '-j', '64']' returned non-zero exit status 2.
```

### Git commit

821189ef6104a16cbfa50e4f911e983b916a3d01

### Host platform

None

### Target Ascend arch (if relevant)

None

### PTOAS build level (if relevant)

None

--- a/lib/PTO/IR/PTO.cpp
+++ b/lib/PTO/IR/PTO.cpp
@@ -6621,10 +6621,22 @@ mlir::LogicalResult mlir::pto::TMrgSortOp::verify() {
       return emitOpError() << "format2 expects dst/tmp element types to match";
     auto dstShape = getShapeVec(dstTy);
     auto tmpShape = getShapeVec(tmpTy);
-    if (dstShape != tmpShape)
-      return emitOpError() << "format2 expects dst/tmp shapes to match";
+    if (dstShape.size() != 2 || tmpShape.size() != 2)
+      return emitOpError() << "format2 expects dst/tmp to be rank-2 tile-shaped";
+    if ((dstShape[0] != mlir::ShapedType::kDynamic && dstShape[0] != 1) ||
+        (tmpShape[0] != mlir::ShapedType::kDynamic && tmpShape[0] != 1))
+      return emitOpError() << "format2 expects dst/tmp rows == 1";
+    if (dstShape[1] != mlir::ShapedType::kDynamic &&
+        tmpShape[1] != mlir::ShapedType::kDynamic &&
+        tmpShape[1] < dstShape[1])
+      return emitOpError() << "format2 expects tmp.cols >= dst.cols";
     for (Value src : getSrcs()) {
       Type srcTy = src.getType();
+      auto srcShape = getShapeVec(srcTy);
+      if (srcShape.size() != 2)
+        return emitOpError() << "format2 expects src to be rank-2 tile-shaped";
+      if (srcShape[0] != mlir::ShapedType::kDynamic && srcShape[0] != 1)
+        return emitOpError() << "format2 expects src rows == 1";
       if (getElemTy(srcTy) != elemTy)
         return emitOpError() << "format2 expects src/dst/tmp element types to match";
     }

--- a/test/lit/pto/tmrgsort_format2_tmp_larger_emitc.pto
+++ b/test/lit/pto/tmrgsort_format2_tmp_larger_emitc.pto
@@ -1,0 +1,14 @@
+// RUN: ptoas --pto-arch=a3 %s 2>&1 | FileCheck %s --check-prefix=A3
+
+module {
+  func.func @tmrgsort_format2_tmp_larger(%ex: vector<4xi16>) {
+    %src0 = pto.alloc_tile : !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=256, v_row=1, v_col=256, blayout=row_major, slayout=none_box, fractal=512, pad=0>
+    %src1 = pto.alloc_tile : !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=256, v_row=1, v_col=256, blayout=row_major, slayout=none_box, fractal=512, pad=0>
+    %dst = pto.alloc_tile : !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=256, v_row=1, v_col=256, blayout=row_major, slayout=none_box, fractal=512, pad=0>
+    %tmp = pto.alloc_tile : !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=512, v_row=1, v_col=512, blayout=row_major, slayout=none_box, fractal=512, pad=0>
+    pto.tmrgsort ins(%src0, %src1, %tmp {exhausted = false} : !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=256, v_row=1, v_col=256, blayout=row_major, slayout=none_box, fractal=512, pad=0>, !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=256, v_row=1, v_col=256, blayout=row_major, slayout=none_box, fractal=512, pad=0>, !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=512, v_row=1, v_col=512, blayout=row_major, slayout=none_box, fractal=512, pad=0>) outs(%dst, %ex : !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=256, v_row=1, v_col=256, blayout=row_major, slayout=none_box, fractal=512, pad=0>, vector<4xi16>)
+    return
+  }
+}
+
+// A3: TMRGSORT<[[DST:Tile<.*256.*>]], [[TMP:Tile<.*512.*>]], [[SRC0:Tile<.*256.*>]], [[SRC1:Tile<.*256.*>]], false>(

--- a/test/lit/pto/tmrgsort_format2_tmp_too_small_invalid.pto
+++ b/test/lit/pto/tmrgsort_format2_tmp_too_small_invalid.pto
@@ -1,0 +1,14 @@
+// RUN: not ptoas --pto-arch=a3 %s 2>&1 | FileCheck %s
+
+module {
+  func.func @tmrgsort_format2_tmp_too_small(%ex: vector<4xi16>) {
+    %src0 = pto.alloc_tile : !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=256, v_row=1, v_col=256, blayout=row_major, slayout=none_box, fractal=512, pad=0>
+    %src1 = pto.alloc_tile : !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=256, v_row=1, v_col=256, blayout=row_major, slayout=none_box, fractal=512, pad=0>
+    %dst = pto.alloc_tile : !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=256, v_row=1, v_col=256, blayout=row_major, slayout=none_box, fractal=512, pad=0>
+    %tmp = pto.alloc_tile : !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=128, v_row=1, v_col=128, blayout=row_major, slayout=none_box, fractal=512, pad=0>
+    pto.tmrgsort ins(%src0, %src1, %tmp {exhausted = false} : !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=256, v_row=1, v_col=256, blayout=row_major, slayout=none_box, fractal=512, pad=0>, !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=256, v_row=1, v_col=256, blayout=row_major, slayout=none_box, fractal=512, pad=0>, !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=128, v_row=1, v_col=128, blayout=row_major, slayout=none_box, fractal=512, pad=0>) outs(%dst, %ex : !pto.tile_buf<loc=vec, dtype=f32, rows=1, cols=256, v_row=1, v_col=256, blayout=row_major, slayout=none_box, fractal=512, pad=0>, vector<4xi16>)
+    return
+  }
+}
+
+// CHECK: error: 'pto.tmrgsort' op format2 expects tmp.cols >= dst.cols

